### PR TITLE
Partially working AD support for BDCs

### PR DESCRIPTION
**Pull request by kevcunnane**
- Plumbed through kerberos support to Notebooks.
  - This works if I replace IP with correct DNS name of "gateway-0" for service, so if this is returned from endpoints table we're all set here
- Plumbed kerberos auth to HDFS, Spark (but this is failing as I haven't implemented correctly yet)
- Fixed error when Data Service node expansion failed and blocked all OE expansion
- Prepped for SqlToolsService change to use new cluster endpoints DMV
  -  Updated API to add new endpoints field to replace IP + port
  - Added logic to handle case where endpoints for Yarn, Grafana etc. are in the list

TODO separately:
- Update backend API as soon as it's ready (cluster I tested against didn't have the DMV needed)
- Handle case where we have IP but it's integrated auth by falling back to static, known service name? May skip this
- Implement kerberos correctly

### CODE REVIEW

File: extensions/mssql/src/dashboard/serviceEndpoints.ts [hunk 1]
```diff
@@ -8,46 +8,69 @@ import * as azdata from 'azdata';
...
+import * as utils from '../utils';
+
+const mgmtProxyName = 'mgmtproxy';
+const grafanaEndpointName = 'metricsui';
+const grafanaDescription = localize('grafana', "Metrics Dashboard");
+const logsuiEndpointName = 'logsui';
+const kibanaDescription = localize('kibana', "Log Search Dashboard");
+const sparkHistoryEndpointName = 'spark-history';
+const sparkHistoryDescription = localize('sparkHistory', "Spark Jobs Management and Monitoring Dashboard");
+const yarnUiEndpointName = 'yarn-ui';
```
Comment (Charles-Gagnon) [inline extensions/mssql/src/dashboard/serviceEndpoints.ts]: We were going to use the description field returned right? Although the old CTP bits don't have that so I suppose if we're supporting that we still need this

File: extensions/mssql/src/dashboard/serviceEndpoints.ts [hunk 2]
```diff
@@ -64,40 +87,57 @@ export function registerServiceEndpoints(context: vscode.ExtensionContext): void
...
+			friendlyName = kibanaDescription;
+			break;
+		case grafanaEndpointName:
+			friendlyName = grafanaDescription;
+			break;
+		case sparkHistoryEndpointName:
+			friendlyName = sparkHistoryDescription;
+			break;
+		case yarnUiEndpointName:
+			friendlyName = yarnHistoryDescription;
```
Comment (Charles-Gagnon) [inline extensions/mssql/src/dashboard/serviceEndpoints.ts]: >yarnHistoryDescription [](start = 18, length = 22)

nit: but I like to consolidate stuff like this and so you could just have getCustomEndpoint call getFriendlyEndpointNames

File: extensions/mssql/src/sparkFeature/sparkUtils.ts [hunk 3]
```diff
@@ -5,13 +5,14 @@
...
 import * as fs from 'fs-extra';
-import * as nls from 'vscode-nls';
 import * as path from 'path';
 import * as azdata from 'azdata';
 import * as vscode from 'vscode';
 import * as which from 'which';
+import * as nls from 'vscode-nls';
 
 const localize = nls.loadMessageBundle();
+import * as constants from '../constants';
```
Comment (Charles-Gagnon) [inline extensions/mssql/src/sparkFeature/sparkUtils.ts]: nit: move with rest of imports

File: resources/xlf/de/mssql.de.xlf [hunk 4]
```diff
@@ -922,7 +922,7 @@
         <source xml:lang="en">Log Search Dashboard</source>
         <target state="translated">Dashboard für Protokollsuche</target>
       </trans-unit>
-      <trans-unit id="sparkHostory">
+      <trans-unit id="sparkhistory">
```
Comment (Charles-Gagnon) [inline resources/xlf/de/mssql.de.xlf]: >sparkhistory [](start = 22, length = 12)

not sure if these are case-sensitive but in the ts file you have sparkHistory

File: extensions/mssql/src/dashboard/serviceEndpoints.ts [hunk 5]
```diff
@@ -8,46 +8,69 @@ import * as azdata from 'azdata';
 import * as nls from 'vscode-nls';
 const localize = nls.loadMessageBundle();
 
-import * as Utils from '../utils';
+import * as utils from '../utils';
+
+const mgmtProxyName = 'mgmtproxy';
+const grafanaEndpointName = 'metricsui';
+const grafanaDescription = localize('grafana', "Metrics Dashboard");
+const logsuiEndpointName = 'logsui';
```
Comment (chlafreniere) [inline extensions/mssql/src/dashboard/serviceEndpoints.ts]: is kibana === logsui? If so, nit: let’s keep the variables endpointname/description in sync

File: extensions/mssql/src/dashboard/serviceEndpoints.ts [hunk 6]
```diff
@@ -8,46 +8,69 @@ import * as azdata from 'azdata';
...
+const kibanaDescription = localize('kibana', "Log Search Dashboard");
+const sparkHistoryEndpointName = 'spark-history';
+const sparkHistoryDescription = localize('sparkHistory', "Spark Jobs Management and Monitoring Dashboard");
```
Comment (chlafreniere) [inline extensions/mssql/src/dashboard/serviceEndpoints.ts]: is it “Jobs” or “Job”

File: extensions/mssql/src/dashboard/serviceEndpoints.ts [hunk 7]
```diff
@@ -8,46 +8,69 @@ import * as azdata from 'azdata';
...
-		});
 		if (endpointsArray.length > 0) {
-			const managementProxyEp = endpointsArray.find(e => e.serviceName === 'management-proxy' || e.serviceName === 'mgmtproxy');
-			if (managementProxyEp) {
-				endpointsArray.push(getCustomEndpoint(managementProxyEp, localize("grafana", "Metrics Dashboard"), '/grafana/d/wZx3OUdmz'));
-				endpointsArray.push(getCustomEndpoint(managementProxyEp, localize("kibana", "Log Search Dashboard"), '/kibana/app/kibana#/discover'));
+			const grafanaEp = endpointsArray.find(e => e.serviceName === grafanaEndpointName);
+			if (grafanaEp) {
+				// Update to have correct URL
+				grafanaEp.endpoint += '/d/wZx3OUdmz';
```
Comment (chlafreniere) [inline extensions/mssql/src/dashboard/serviceEndpoints.ts]: 🤐

File: extensions/mssql/src/dashboard/serviceEndpoints.ts [hunk 8]
```diff
@@ -8,46 +8,69 @@ import * as azdata from 'azdata';
...
-			const gatewayEp = endpointsArray.find(e => e.serviceName === 'gateway');
-			if (gatewayEp) {
-				endpointsArray.push(getCustomEndpoint(gatewayEp, localize("sparkHostory", "Spark Job Monitoring"), '/gateway/default/sparkhistory'));
-				endpointsArray.push(getCustomEndpoint(gatewayEp, localize("yarnHistory", "Spark Resource Management"), '/gateway/default/yarn'));
+			if (!grafanaEp) {
+				// We are on older CTP, need to manually add some endpoints.
+				// TODO remove once CTP support goes away
+				const managementProxyEp = endpointsArray.find(e => e.serviceName === mgmtProxyName);
+				if (managementProxyEp) {
+					endpointsArray.push(getCustomEndpoint(managementProxyEp, grafanaEndpointName, grafanaDescription, '/grafana/d/wZx3OUdmz'));
```
Comment (chlafreniere) [inline extensions/mssql/src/dashboard/serviceEndpoints.ts]: we’re adding this in a couple of places. would be great to have this list in a constant somewhere.

File: extensions/mssql/src/objectExplorerNodeProvider/objectExplorerNodeProvider.ts [hunk 9]
```diff
@@ -325,15 +325,20 @@ class DataServicesNode extends TreeNode {
...
-			this._children.push(hdfsNode);
+			return this.refreshChildren();
 		}
 		return this._children;
 	}
 
+	private async refreshChildren(): Promise<TreeNode[]> {
+		this._children = [];
+		let fileSource: IFileSource = await this.session.sqlClusterConnection.createHdfsFileSource();
+		let hdfsNode = new ConnectionNode(this._context, localize('hdfsFolder', 'HDFS'), fileSource);
```
Comment (chlafreniere) [inline extensions/mssql/src/objectExplorerNodeProvider/objectExplorerNodeProvider.ts]: double quotes 😄

File: extensions/mssql/src/sparkFeature/dialog/sparkJobSubmission/sparkJobSubmissionService.ts [hunk 10]
```diff
@@ -160,7 +176,8 @@ export class SparkJobSubmissionInput {
 		private _port?: number,
 		private _livyPath?: string,
 		private _user?: string,
-		private _passWord?: string) {
+		private _passWord?: string,
```
Comment (chlafreniere) [inline extensions/mssql/src/sparkFeature/dialog/sparkJobSubmission/sparkJobSubmissionService.ts]: nit: should the ‘W’ be capitalized

File: extensions/mssql/src/util/auth.ts [hunk 11]
```diff
@@ -0,0 +1,14 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as kerberos from 'kerberos';
+
+export async function authenticateKerberos(hostname: string): Promise<string> {
+	const service = `HTTP@${hostname}`;
```
Comment (chlafreniere) [inline extensions/mssql/src/util/auth.ts]: dumb question should this be HTTP or HTTPS

File: src/sql/workbench/services/objectExplorer/common/objectExplorerService.ts [hunk 12]
```diff
@@ -124,6 +124,8 @@ export interface TopLevelChildrenPath {
 	providerObject: azdata.ObjectExplorerNodeProvider | azdata.ObjectExplorerProvider;
 }
 
+const errSessionCreateFailed = nls.localize('OeSessionFailedError', "Failed to create Object Explorer session");
```
Comment (chlafreniere) [inline src/sql/workbench/services/objectExplorer/common/objectExplorerService.ts]: Oe looks weird to me. But it’s fine.

Comment (kevcunnane) [reply src/sql/workbench/services/objectExplorer/common/objectExplorerService.ts]: I mean I'm using what already was there, just moving it.

**Pull request merged**